core/src/main/scala/fs2/concurrent.scala

```diff
@@ -1,17 +1,38 @@
 package fs2
 
-import util.Free
 
 object concurrent {
 
-  def join[F[_],O](maxOpen: Int)(s: Stream[F,Stream[F,O]])(implicit F: Async[F]): Stream[F,O] = {
-    if (maxOpen <= 0) throw new IllegalArgumentException("maxOpen must be > 0, was: " + maxOpen)
+  /**
+    * Joins non deterministically streams.
+    *
+    *
+    * @param maxOpen    Allows to specify maximum open streams at any given time.
+    *                   The outer stream will stop its evaluation when this limit is reached and will
+    *                   start evaluating once currently open streams will be <= mayOpen.
+    *                   MaxOpen must be > 0
+    *
+    * @param outer      Outer stream, that produces streams (inner) to be run concurrently.
+    *                   When this stops gracefully, then all inner streams are continuing to run
+    *                   resulting in process that will stop when all `inner` streams finished
+    *                   their evaluation.
+    *
+    *                   When this stream fails, then evaluation of all `inner` streams is interrupted
+    *                   and resulting stream will fail with same failure.
+    *
+    *                   When any of `inner` streams fails, then this stream is interrupted and
+    *                   all other `inner` streams are interrupted as well, resulting in stream that fails
+    *                   with error of the stream that cased initial failure.
+    */
+  // todo: document when the finalizers are calle in which situation
+  def join[F[_],O](maxOpen: Int)(outer: Stream[F,Stream[F,O]])(implicit F: Async[F]): Stream[F,O] = {
+    assert(maxOpen > 0,"maxOpen must be > 0, was: " + maxOpen)
     for {
       killSignal <- Stream.eval(async.signalOf(false))
       openLeases <- Stream.eval(async.mutable.Semaphore(maxOpen))
       outerDone <- Stream.eval(F.refOf[Boolean](false))
       outputQueue <- Stream.eval(async.mutable.Queue.synchronousNoneTerminated[F,Either[Throwable,Chunk[O]]])
-      o <- s.evalMap { (s: Stream[F,O]) =>
+      o <- outer.evalMap { (inner: Stream[F,O]) =>
```

Comment:
sorry to touch you code this way:-) I just wanted to be sure I understand what is called where, and didn't revert it. You want me to?